### PR TITLE
[IBFT] Change quorum size calculation

### PR DESCRIPTION
**Pull request by dbrajovic**
# Description

This PR replaces the old calculation of finding the minimum number of validators reaching quorum.

`N` - number of validators in network
`F = floor((N-1) / 3)` - number of (allowed) faulty nodes in a given network

Previously :

`quorum = 2 * F + 1`

Now:

`quorum = ceil(2/3 * N)`

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [ ] I have tested this code manually

### Manual tests

Please complete this section if you ran manual tests for this functionality, otherwise delete it

### Additional comments
Fixes EDGE-514

The formula (previously used) for calculating the Quorum size, which is `2F + 1` did not scale well with the number of validators in a given network. Note the following differences:

`2F + 1`:

- `N = 1`, `F = 0` -> `Q = 1`
- `N = 2`, `F = 0` -> `Q = 1`
- `N = 3`, `F = 0` -> `Q = 1`
- `N = 4`, `F = 1` -> `Q = 3`
- `N = 5`, `F = 1` -> `Q = 3`
- `N = 6`, `F = 1` -> `Q = 3`
- `N = 7`, `F = 2` -> `Q = 5`
- `N = 8`, `F = 2` -> `Q = 5`
- `N = 9`, `F = 2` -> `Q = 5`

`ceil(2/3 * N)`:
 from [table](https://www.kaleido.io/blockchain-blog/consensus-algorithms-poa-ibft-or-raft#:~:text=Istanbul%20Byzantine%20Fault%20Tolerance%20(IBFT)) (for `N=3` the formula does not apply, so a special [if statement](https://github.com/0xPolygon/polygon-edge/blob/44c154be5602fe978d70f9f64ad40a84ccf257d6/consensus/ibft/state.go#L311) was used)

- `N = 1`, `F = 0` -> `Q = 1`
- `N = 2`, `F = 0` -> `Q = 2`
- `N = 3`, `F = 0` -> `Q = 3`
- `N = 4`, `F = 1` -> `Q = 3`
- `N = 5`, `F = 1` -> `Q = 4`
- `N = 6`, `F = 1` -> `Q = 4`
- `N = 7`, `F = 2` -> `Q = 5`
- `N = 8`, `F = 2` -> `Q = 6`
- `N = 9`, `F = 2` -> `Q = 6`

### CODE REVIEW

**Comment by Kourin1996** [comment]
@dbrajovic Can you write down the reason so that we can look back in the future?

File: consensus/ibft/state.go [hunk 1]
```diff
@@ -97,12 +98,15 @@ func (c *currentState) setState(s IbftState) {
...
-	// (the current node tallying the messages will include its own message)
-	return 2 * c.validators.MaxFaultyNodes()
+func NumValid(validators ValidatorSet) int {
+	//	if the number of validators is less than 4,
+	//	then the entire set is required
+	if validators.MaxFaultyNodes() == 0 {
+		return validators.Len()
+	}
+
+	// 	Q = ceil(2/3 * N) - (quorum optimal)
```
Comment (Kourin1996) [inline consensus/ibft/state.go]: What does `- (quorum optimal)` mean?

File: consensus/ibft/state.go [hunk 2]
```diff
@@ -97,12 +98,15 @@ func (c *currentState) setState(s IbftState) {
 }
 
 // NumValid returns the number of required messages
-func (c *currentState) NumValid() int {
-	// According to the IBFT spec, the number of valid messages
-	// needs to be 2F + 1
-	// The 1 missing from this equation is accounted for elsewhere
-	// (the current node tallying the messages will include its own message)
-	return 2 * c.validators.MaxFaultyNodes()
+func NumValid(validators ValidatorSet) int {
```
Comment (zivkovicmilos) [inline consensus/ibft/state.go]: What do you think about renaming this to `QuorumSize`?

Comment (Kourin1996) [inline consensus/ibft/state.go]: Why did you change from method to function? I think currentState or ValidattorSet should have this

**Review by zivkovicmilos — APPROVED** [review]
Looks good 💯 

I've went through the code and the spec, along with Saltini's paper, the logic seems to be sound.
I don't like that the formula differs for `N < 4`, but 🤷‍♂️ .

File: consensus/ibft/state.go (hunk 2 above)
Comment (Kourin1996) [reply consensus/ibft/state.go]: The reason why I thought is that you can't change calculations without rewriting the code. You just create new struct and method when you want to change calc if you define it as method

File: consensus/ibft/state.go (hunk 1 above)
Comment (dbrajovic) [reply consensus/ibft/state.go]: Misleading comment

[Fixed](https://github.com/0xPolygon/polygon-edge/pull/513/commits/058e3c6d5fd3b6afe0b09ef981dae87a0441c95d)

File: consensus/ibft/state.go (hunk 2 above)
Comment (dbrajovic) [reply consensus/ibft/state.go]: [Renamed](https://github.com/0xPolygon/polygon-edge/pull/513/commits/6e5f031b9e2f50cc8129606f41001c194b87c505)

Comment (dbrajovic) [reply consensus/ibft/state.go]: Good point 👍 

[Refactored](https://github.com/0xPolygon/polygon-edge/pull/513/commits/44c154be5602fe978d70f9f64ad40a84ccf257d6)

**Review by Kourin1996 — APPROVED** [review]
LGTM!

File: txpool/txpool.go [hunk 3]
```diff
@@ -564,6 +564,7 @@ func (p *TxPool) addTx(origin txOrigin, tx *types.Transaction) error {
 
 	// check if already known
 	if _, ok := p.index.get(tx.Hash); ok {
+		fmt.Println("known tx", "hash", tx.Hash.String())
```
Comment (0xAleksaOpacic) [inline txpool/txpool.go]: why do we need this log

**Review by 0xAleksaOpacic — APPROVED** [review]

File: txpool/txpool.go (hunk 3 above)
Comment (dbrajovic) [reply txpool/txpool.go]: Debug print for the GH test runs (got what I needed)

[Removed](https://github.com/0xPolygon/polygon-edge/pull/513/commits/ce66280be29bcea7c9e73f297981df266ec3910b)
